### DISCUSSION TITLE
adding PCSC support to nfcgate

### DISCUSSION THREAD

**Discussion by 0dayfrp** [nfcgate/server#8] [Q&A]
Hello, I'm trying to communicate from windows pc with Nfcgate server, I have not decided yet which is the best approach. 
Can you please advise? below is an attempt I have made 

import c2c_pb2.py, c2s_pb2.py
import socket
import socketserver
import struct
import sys
import datetime
from smartcard.System import readers
from smartcard.Exceptions import NoCardException

HOST = "0.0.0.0"
PORT = 5566

class PluginHandler:
    # Existing PluginHandler code remains unchanged

class NFCGateClientHandler(socketserver.StreamRequestHandler):
    # Existing __init__, log, and setup methods remain unchanged

    def communicate_with_card(self, apdu):
        r = readers()
        if len(r) == 0:
            self.log("No smart card readers detected.", tag="SmartCard")
            return None

        try:
            connection = r[0].createConnection()
            connection.connect()
            response, sw1, sw2 = connection.transmit(apdu)
            return response + [sw1, sw2]
        except NoCardException:
            self.log("No smart card inserted.", tag="SmartCard")
            return None

    def handle(self):
        # Existing handle setup code remains unchanged

        while True:
            try:
                # Existing code to read data from the socket remains unchanged

                # Here, integrate smart card communication
                # Convert data to APDU format if necessary
                apdu = list(data)  # Ensure it's in the correct format for pyscard
                card_response = self.communicate_with_card(apdu)
                
                if card_response is not None:
                    self.log(f"Card response: {card_response}", tag="SmartCard")
                    # Modify this part to send the card's response back
                    # For example, you might send it directly back to the requester or to all clients in a session
                    # This is a placeholder; adapt it to your needs
                    self.wfile.write(struct.pack('!I', len(card_response)) + bytes(card_response))
                else:
                    self.log("Error communicating with the card.", tag="SmartCard")

                # Break or continue based on your application's logic

            except socket.timeout:
                # Existing timeout handling remains unchanged

    # Existing finish method remains unchanged

class NFCGateServer(socketserver.ThreadingTCPServer):
    # Existing NFCGateServer code remains unchanged

if __name__ == "__main__":
    NFCGateServer((HOST, PORT), NFCGateClientHandler).serve_forever()

**Comment by 0dayfrp** [comment]
sorry for posting in wrong category.
Another attempt on Windows 11 Broadcom Corp Contacted Smartcard 0 
Broadcom Corp Contactless Smartcard 0  pyscard working, code connects to server but does not join session.

loop.py

import socket
import select
from smartcard.System import readers
from smartcard.Exceptions import NoCardException
import struct
import time

# NFCGate server details
SERVER_HOST = '192.168.3.74'
SERVER_PORT = 5566
SESSION_ID = 11  # Default session ID

def connect_to_reader():
    """Connect to the PC/SC reader."""
    reader_list = readers()
    if not reader_list:
        print("No PC/SC reader detected.")
        return None

    reader = reader_list[0]
    print(f"Connecting to reader: {reader}")

    try:
        connection = reader.createConnection()
        connection.connect()
        return connection
    except NoCardException:
        print("No card detected in the reader.")
        return None

def communicate_with_server(card_connection):
    """Handle bidirectional communication with the NFCGate server."""
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        sock.connect((SERVER_HOST, SERVER_PORT))
        sock.setblocking(0)
        print("Connected to NFCGate server.")

        try:
            while True:
                # Example: Polling for data from the card and server
                ready_to_read, _, _ = select.select([sock], [], [], 0.5)

                # Check for incoming data from server
                for s in ready_to_read:
                    server_data = s.recv(1024)
                    if server_data:
                        # Process and possibly relay data back to the card
                        print(f"Received from server: {server_data}")
                        # Add logic here to send data back to the card if necessary

                # Example: Sending data to the server (simulate card activity)
                # Here you would read from the card and send data to the server
                # This is just an example and should be replaced with actual card reading logic
                # command_apdu = ...
                # response_apdu = card_connection.transmit(command_apdu)
                # response_bytes = bytes(response_apdu)
                # message = struct.pack('!IB', len(response_bytes), SESSION_ID) + response_bytes
                # sock.sendall(message)

                # Placeholder for sleep to simulate delay between card interactions
                # Remove or adjust this based on actual card interaction logic
                time.sleep(1)

        except Exception as e:
            print(f"Error during communication: {e}")

def main():
    card_connection = connect_to_reader()
    if card_connection:
        communicate_with_server(card_connection)

if __name__ == "__main__":
    main()



and this is server.py


#!/usr/bin/env python3
import socket
import socketserver
import struct
import sys
import datetime

HOST = "192.168.3.74"
PORT = 5566


class PluginHandler:
    def __init__(self):
        self.plugin_list = []

        for modname in sys.argv[1:]:
            self.plugin_list.append((modname, __import__("plugins.mod_%s" % modname, fromlist=["plugins"])))
            print("Loaded", "mod_%s" % modname)

    def filter(self, client, data):
        for modname, plugin in self.plugin_list:
            if type(data) == list:
                first = data[0]
            else:
                first = data
            first = plugin.handle_data(lambda *x: client.log(*x, tag=modname), first, client.state)
            if type(data) == list:
                data = [first] + data[1:]
            else:
                data = first

        return data


class NFCGateClientHandler(socketserver.StreamRequestHandler):
    def __init__(self, request, client_address, srv):
        super().__init__(request, client_address, srv)
        
    def log(self, *args, tag="server"):
        self.server.log(*args, origin=self.client_address, tag=tag)

    def setup(self):
        super().setup()
        
        self.session = 11
        self.state = {}
        self.request.settimeout(900)
        self.log("server", "connected")

    def handle(self):
        super().handle()

        while True:
            try:
                msg_len_data = self.rfile.read(5)
            except socket.timeout:
                self.log("server", "Timeout")
                break
            if len(msg_len_data) < 5:
                break

            msg_len, session = struct.unpack("!IB", msg_len_data)
            data = self.rfile.read(msg_len)
            self.log("server", "data:", bytes(data))

            # no data was sent or no session number supplied and none set yet
            if msg_len == 0 or session == 0 and self.session is None:
                break

            # change in session number detected
            if self.session != session:
                # remove from old association
                self.server.remove_client(self, self.session)
                # update and add association
                self.session = session
                self.server.add_client(self, session)

            # allow plugins to filter data before sending it to all clients in the session
            self.server.send_to_clients(self.session, self.server.plugins.filter(self, data), self)

    def finish(self):
        super().finish()

        self.server.remove_client(self, self.session)
        self.log("server", "disconnected")


class NFCGateServer(socketserver.ThreadingTCPServer):
    def __init__(self, server_address, request_handler, bind_and_activate=True):
        self.allow_reuse_address = True
        super().__init__(server_address, request_handler, bind_and_activate)

        self.clients = {}
        self.plugins = PluginHandler()
        self.log("NFCGate server listening on", server_address)
        
    def log(self, *args, origin="0", tag="server"):
        print(datetime.datetime.now(), "["+tag+"]", origin, *args)

    def add_client(self, client, session):
        if session is None:
            return

        if session not in self.clients:
            self.clients[session] = []

        self.clients[session].append(client)
        client.log("joined session", session)

    def remove_client(self, client, session):
        if session is None or session not in self.clients:
            return

        self.clients[session].remove(client)
        client.log("left session", session)

    def send_to_clients(self, session, msgs, origin):
        if session is None or session not in self.clients:
            return

        for client in self.clients[session]:
            # do not send message back to originator
            if client is origin:
                continue

            if type(msgs) != list:
                msgs = [msgs]

            for msg in msgs:
                client.wfile.write(int.to_bytes(len(msg), 4, byteorder='big'))
                client.wfile.write(msg)

        self.log("Publish reached", len(self.clients[session]) - 1, "clients")


if __name__ == "__main__":
    NFCGateServer((HOST, PORT), NFCGateClientHandler).serve_forever()

**Comment by 0dayfrp** [comment]
@kleest  any ideas on next step? 

Thank you for your good work.

**Reply by kleest** [reply]
Please format the code so that it is readable and syntax highlighting works correctly.

**Reply by 0dayfrp** [reply]
Hello thanks for reply I'm just a beginner but learn very fast please explain like for children :-)
[loop.zip](https://github.com/nfcgate/server/files/14233294/loop.zip)

**Reply by 0dayfrp** [reply]
[1sttry.zip](https://github.com/nfcgate/server/files/14233344/1sttry.zip)
my first attempt

**Reply by 0dayfrp** [reply]
@Mydayyy this may seem to you out of this world because it is (I'm a novice at this but not stupid) but instead of just reacting maybe a little guidance?  Love u 2

**Comment by roussosalex** [comment]
@0dayfrp This is a very interesting concept. Unfortunately I will have to close this discussion as it is attracting low-quality interactions. Please open a new discussion with the following improvements:
- provide more detail on what exactly you are trying to do, *how* it is not working, *what* is the error exactly, etc.
- provide only *relevant* portions of your code, do not just post everything
- do not post files you slightly edited completely, instead write how and what you edited. Ideally provide a "git diff" of such files.
- take a minute to properly format your code so we can read it ([tutorial](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
